text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,237 @@
+This proposal introduces extern types, a way to accurately represent these opaque types in the Rust type system such that the compiler prevents you from misusing it.
+
+The motivation for replacing [RFC 1861] is around computing the offset of fields in aggregate types.
+Extern types do not have an alignment known to the Rust compiler and therefore cannot be included as a field in an aggregate type as it is impossible to correctly calculate their offset, however that RFC did not address this.
+This implies that extern types cannot be included in any (non-`repr(transparent)`) structs and so must be prevented.
+
+Extern types should be able to be used in generic contexts so that they receive blanket trait implementations and can be used inside generic wrapper types.
+Prior to this, all generic types could be included as fields of structs and it was possible to obtain the size and alignment of any type.
+Extern types do not have a size or alignment so this RFC allows users to specify generic bounds sufficiently to statically prevent the above issues.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing bindings to foreign code, any types that are opaque to the Rust type system should be declared as follows:
+```rust
+extern {
+    type Foo;
+}
+```
+This is called an extern type and is `!Sized`, `!MetaSized` (explained below), `!Send`, `!Sync`, and is FFI-safe.
+Unlike other dynamically-sized types (DSTs), currently only slices and trait objects, pointers to it (`&Foo`, `&mut Foo`, `*const Foo`, `*mut Foo`, etc) are thin, that is they are 1 `usize` wide not 2.
+
+These types cannot be included in structs as Rust is not able to compute the offset of the field because it does not know the alignment.
+```rust
+struct Header {
+    data: u8,
+    tail: Foo, // Error due to unknown offset
+}
+```
+
+However, these types may be placed in `repr(transparent)` structs:
+```rust
+#[repr(transparent)]
+struct Wrapper<T> {
+    inner: Foo,
+    _marker: PhantomData<T>,
+}
+```
+
+The lack of the `Sized` and `MetaSized` traits on these structs prevent you from calling `ptr::read`, `mem::size_of_val`, etc, which are not meaningful for opaque types.
+
+In the 2021 edition and earlier, these types cannot be used in generic contexts as `T: Sized` and `T: ?Sized` both imply that `T` has a computable size and alignment.
+
+In the 2024 edition and later, `T: ?Sized` no longer implies any knowledge of the size and alignment so opaque types can be used in generic contexts.
+If you require your generic type to have a computable size and alignment you can use the bound `T: ?Sized + MetaSized`, which will enable you to store the type in a struct.
+
+The automated tooling for migrating from the 2021 edition to the 2024 edition will replace `?Sized` bounds with `?Sized + MetaSized` bounds.
```

Comment:
The tooling should only do the replacement if `Metasized` is required, or a lot of declarations will get a lot noisier.

I feel it would be useful to have an idea on how much of an impact this will have on the ecosystem -- how common will it be to need a `Metasized` bound.

Doing the review of the standard library mentioned later would be a start.

Reply:
> The tooling should only do the replacement if `Metasized` is required, or a lot of declarations will get a lot noisier.

How do you propose deciding which case is which? Note that (usually) relaxing the bound is non-breaking but adding it in later is breaking so a human decision is often required for what promises the API wants to make.